url2df0f7c811_przemekgalazka_legacycode_tgz-pJ8-pl_cc_core_cmd_CommandJ8/src/ccproxy/QueueMemberList.java

```diff
@@ -0,0 +1,97 @@
+package ccproxy;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class QueueMemberList {
+	List<QueueMember> aclist = Collections.synchronizedList( new ArrayList<QueueMember>() );
+
+	public QueueMemberList() {
+
+	}
+
+	public synchronized int size() {
+		return aclist.size();
+	}
+
+	public synchronized QueueMember get(int i) {
+		return aclist.get(i);
+	}
+
+	public synchronized boolean contains(Object o) {
+		return aclist.contains(o);
+	}
+
+	public synchronized boolean containsNumer(String s) {
+		for (int i = 0; i < aclist.size(); i++) {
+			QueueMember qm = aclist.get(i);
+			if (qm.ac.agentNumber.equals(s))
+				return true;
+		}
+		return false;
+	}
+
+	public synchronized void add(QueueMember ac) {
+		aclist.add(ac);
+	}
+
+	public synchronized void remove(QueueMember ac) {
+		aclist.remove(ac);
+	}
+
+	public synchronized void remove(int i) {
+		aclist.remove(i);
+	}
+
+	public synchronized QueueMember getByIP(String ip) {
+		for (int i = 0; i < aclist.size(); i++) {
+			if (aclist.get(i).ac.ip.equals(ip))
+				return aclist.get(i);
+		}
+		return null;
+	}
+
+	public synchronized QueueMember getQueueMemberByNumer(String number) {
+		for (int i = 0; i < aclist.size(); i++) {
+			if (aclist.get(i).ac.agentNumber.equals(number))
+				// //System.out.println("dupadapduapdsa");
+				return aclist.get(i);
+		}
+		return null;
+	}
+
+	public synchronized void pushMessage(String message) {
+		for (int i = 0; i < aclist.size(); i++) {
+			aclist.get(i).ac.out.println(message);
+		}
+	}
+
+	public void pushMessage(String message, QueueMember qm) {
+		qm.ac.out.println(message);
+	}
+
+	public void pushQueueMessage(String message) {
+		for (int i = 0; i < aclist.size(); i++) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **75**. At this line the following field is not protected by synchronization: `aclist`. Your code typically protects this field with synchronization on object: `QueueMemberList.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `QueueMemberList.this`. For example, like at lines: **15**, **19**, **40**, **44**, **36**, **23**, **65**, **85**, **86**, and **66**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **15**, **19**, **23**, **27**, **28**, **36**, **40**, **44**, **48**, **49**, **50**, **56**, **57**, **59**, **65**, **66**, **85**, and **86**. This is a call to `size()`. You are already protecting `size()`. You are protecting these operations at lines: **15**.